pkg/controller/pipelinerun/controller.go

```diff
@@ -0,0 +1,241 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pipelinerun
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
+	clientset "github.com/knative/build-pipeline/pkg/client/clientset/versioned"
+	"github.com/knative/build-pipeline/pkg/client/clientset/versioned/scheme"
+	informers "github.com/knative/build-pipeline/pkg/client/informers/externalversions"
+	pi "github.com/knative/build-pipeline/pkg/client/informers/externalversions/pipeline/v1alpha1"
+	listers "github.com/knative/build-pipeline/pkg/client/listers/pipeline/v1alpha1"
+	"github.com/knative/build-pipeline/pkg/constants"
+	"github.com/knative/build-pipeline/pkg/controller"
+	"github.com/knative/pkg/logging/logkey"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// for testing
+var (
+	runsListerFunc      = runsLister
+	pipelinesListerFunc = pipelinesLister
+	getRecordFunc       = getRecorder
```

Comment:
instead of having these as globals, how about having two `NewController` functions, one which takes these functions as params, which the test could use, and one which hardcodes them?

i kind of like that that would limit the test specific code to one function, and wouldn't involve any globals